### ISSUE TITLE
UTF-8 Bug wih Monthnames in German language

### ISSUE THREAD

**Issue by wintstar** [SimpleMachines/SMF#4728]
Version Information:
Forum version: SMF 2.1 Beta 4 (more detailed)
Current SMF version: SMF 2.1 Beta 3
GD version: bundled (2.1.0 compatible)
MySQLi System: MySQL 
MySQLi Version: 5.6.39-log 
Alternative PHP Cache: 5.1.2
PHP: 7.1.11 (more detailed)
Server version: Apache/2.4.29 (Win32) OpenSSL/1.0.2l PHP/7.1.11
Browser: Chrome Version 65.0.3325.181

Pull Request # 4714  from 15.04.2018 19:56

With written out month names there is a bug with umlauts

My index.german.php. You can download the complete German language package on my 
website.

```
// Version: 2.1 Beta 4; index

global $forum_copyright, $webmaster_email, $scripturl, $context, $modSettings, $boardurl;

// Native name, please use full HTML entities to write your language's name.
$txt['native_name'] = 'Deutsch';

// Locale (strftime, pspell_new) and spelling. (pspell_new, can be left as '' normally.)
// For more information see:
//   - https://php.net/function.pspell-new
//   - https://php.net/function.setlocale
// Again, SPELLING SHOULD BE '' 99% OF THE TIME!!  Please read this!
$txt['lang_locale'] = 'de_DE';
$txt['lang_dictionary'] = 'de';
$txt['lang_spelling'] = 'german';
//https://developers.google.com/recaptcha/docs/language
$txt['lang_recaptcha'] = 'de';

// Ensure you remember to use uppercase for character set strings.
$txt['lang_character_set'] = 'UTF-8';
// Character set and right to left?
$txt['lang_rtl'] = false;
// Number format.
$txt['number_format'] = '1.234.00';
```
![board-umlauts-bug](https://user-images.githubusercontent.com/2753863/39268504-6738c9ac-48d0-11e8-9d28-529c0850912b.PNG)

![stats-umlauts-bug](https://user-images.githubusercontent.com/2753863/39268511-6c67ca2c-48d0-11e8-9524-0d8e0016b871.PNG)

![index-umlauts bug](https://user-images.githubusercontent.com/2753863/39268516-73692cf8-48d0-11e8-8f51-f006b33b1e52.PNG)

**Comment by albertlast** [comment]
I download your packages and
use short and long month display format but the result is always the same:
![grafik](https://user-images.githubusercontent.com/1782906/39315089-24f65054-4976-11e8-8db2-1969ad8e4428.png)
That i got not this issue.
Testet with mysql and pg.

**Comment by wintstar** [comment]
@albertlast , and your php Version?

I have the problem only with the test forum 2.1 beta 4.  With version SMF 2.0.15, installed on the same server, I do not have this problem.

**Comment by albertlast** [comment]
7.2

**Comment by wintstar** [comment]
PHP: 7.1.11

**Comment by albertlast** [comment]
The issue is in my eyes not related to the php version...

**Comment by wintstar** [comment]
What could be the reason, as previously written 2.0.15 no problem, 2.1 beta 4 is the problem. Also under xampp / localhost. The problem only occurs with month names, nowhere else.

**Comment by albertlast** [comment]
i would guess that your language file got corrupt,
replace them with your packages files.

**Comment by wintstar** [comment]
> i would guess that your language file got corrupt,
> replace them with your packages files.

You tested my package and it was fine with you. So I rule that out.

**Comment by albertlast** [comment]
I'm not sure what you mean,
but yes i used your packages with xampp and
like the screenshot shows, it work fine.

**Comment by wintstar** [comment]
You have write "i would guess that your language file got corrupt" 

> German language
>Du hast geschrieben das mein Sprachpaket fehlerhaft ist. War auch meine erste Vermutung. Habe dann das Sprachpaket ausgetauscht. Aber das Problem ist immer noch da.

You wrote that my language pack was corrupt. That was my first guess, too. Then I exchanged the language pack. But the problem is still there.
I also exchanged the language files again, no change

**Comment by albertlast** [comment]
Nope your language pack is fine like my screenshot show.
You could zip and upload the html code of one of the pages where the error appears.

**Comment by wintstar** [comment]
[utf-8-bug.zip](https://github.com/SimpleMachines/SMF2.1/files/1952777/utf-8-bug.zip)

But it is not only in the board overview

**Comment by albertlast** [comment]
I want the generated html code which is shown in your browser -> go to your 2.1 setup with you browser than save the page on your local storage.

**Comment by wintstar** [comment]
[boardindex.zip](https://github.com/SimpleMachines/SMF2.1/files/1952803/boardindex.zip)

You have a PM on simplemachines.org

**Comment by albertlast** [comment]
For comparsion would be nice to get your index.german.php (i know that they is included in your packages) but i would like to see if some different exists in your version which you use on your webserver.

**Comment by wintstar** [comment]
[index.german.php.zip](https://github.com/SimpleMachines/SMF2.1/files/1952910/index.german.php.zip)

**Comment by albertlast** [comment]
When you look at your smf2.1 env, in the stats area,
you see that your month there are written right,
so the problem is not easy for me to get into...

When you look at response of you page then you see that it look fine for the month but for the other umlauts it look wrong:
![grafik](https://user-images.githubusercontent.com/1782906/39331369-a31b5bfe-49a4-11e8-88f9-a570be95b877.png)

So in general i keep on point that the issue is on your webserver side,
but i could not say what the issue is.

**Comment by wintstar** [comment]
> So in general i keep on point that the issue is on your webserver side,
> but i could not say what the issue is.

That's my guess, too. I have been troubleshooting for a long time before I created this issue. I didn't find anything. And since this problem is not in the 2.0.15 installation (itself server), I suspect the problem somewhere in 2.1 beta4. and surprisingly it only affects the month name.

**Comment by albertlast** [comment]
When you change in Firefox (in Chrome this feature doesn't exists) the charset from utf-8 to windows 1252 you see,
that the month are displayed correctly but the other get wrong (because they are in utf-8).

**Comment by albertlast** [comment]
Could you change in your local index.geman.php 
```
$txt['months'] = array(1 => 'Januar', 'Februar', 'März', 'April', 'Mai', 'Juni', 'Juli', 'August', 'September', 'Oktober', 'November', 'Dezember');
$txt['months_titles'] = array(1 => 'Januar', 'Februar', 'März', 'April', 'Mai', 'Juni', 'Juli', 'August', 'September', 'Oktober', 'November', 'Dezember');
$txt['months_short'] = array(1 => 'Jan', 'Feb', 'Mär', 'Apr', 'Mai', 'Jun', 'Jul', 'Aug', 'Sep', 'Okt', 'Nov', 'Dez');
```
to
```
$txt['months'] = array(1 => 'Januar', 'Februar', 'Märze', 'April', 'Mai', 'Juni', 'Juli', 'August', 'September', 'Oktober', 'November', 'Dezember');
$txt['months_titles'] = array(1 => 'Januar', 'Februar', 'Märze', 'April', 'Mai', 'Juni', 'Juli', 'August', 'September', 'Oktober', 'November', 'Dezember');
$txt['months_short'] = array(1 => 'Jan', 'Feb', 'Mäo', 'Apr', 'Mai', 'Jun', 'Jul', 'Aug', 'Sep', 'Okt', 'Nov', 'Dez');
```
And look if this changes affect your page displayed month(März should look now different)?

**Comment by wintstar** [comment]
I have already tested it, but the changes are not displayed despite empty Chache. I just changed the language file with your default. I have installed a new test forum on xampp, there is no problem. So the error come with update (upgrade.php) must have happened. Only what could have been updated wrongly.

**Comment by albertlast** [comment]
I don't understand you awnser,
when you change the file,
get März displayed differently(should be now Märze) or not?

**Comment by wintstar** [comment]
Have called the test forum now with Firefox. Cache manually emptied on the server and in the browser. The changes to the language file are not displayed. Even if I change the English index.english.php the text for the month names to German.

**Comment by wintstar** [comment]
Yes, I changed the language file, should now show Maerze. Even if you now switch to English as your language.

> German
> Ja ich habe die Sprachdatei geändert, sollte jetzt Maerze angezeigt werden. Auch wenn du jetzt auf Englisch als Sprache umstellst

**Comment by wintstar** [comment]
I have send a PM for you.

**Comment by albertlast** [comment]
I will check this in the afternoon.


________________________________
Von: wintstar <notifications@github.com>
Gesendet: Freitag, 27. April 2018 06:57
An: SimpleMachines/SMF2.1
Cc: albertlast; Mention
Betreff: Re: [SimpleMachines/SMF2.1] UTF-8 Bug wih Monthnames in German language (#4728)


I have send a PM for you.

**Comment by wintstar** [comment]
Thank you. I have created a database update. You can test everything.

**Comment by wintstar** [comment]
Could it be the database?

xampp, no bug
MySQLi Version: 10.1.29-MariaDB 

my Testboard, there's the mistake.
MySQLi Version: 5.6.39-log

**Comment by albertlast** [comment]
Could you look into your smf_settings table which value the key:  global_character_set got

**Comment by wintstar** [comment]
```
<?php

/**
 * The settings file contains all of the basic settings that need to be present when a database/cache is not available.
 *
 * Simple Machines Forum (SMF)
 *
 * @package SMF
 * @author Simple Machines http://www.simplemachines.org
 * @copyright 2018 Simple Machines and individual contributors
 * @license http://www.simplemachines.org/about/smf/license.php BSD
 *
 * @version 2.1 Beta 4
 */

########## Maintenance ##########
/**
 * The maintenance "mode"
 * Set to 1 to enable Maintenance Mode, 2 to make the forum untouchable. (you'll have to make it 0 again manually!)
 * 0 is default and disables maintenance mode.
 * @var int 0, 1, 2
 * @global int $maintenance
 */
$maintenance = 0;
/**
 * Title for the Maintenance Mode message.
 * @var string
 * @global int $mtitle
 */
$mtitle = 'Wartungsmodus';
/**
 * Description of why the forum is in maintenance mode.
 * @var string
 * @global string $mmessage
 */
$mmessage = 'Es wird ein neues Pull Request installiert ... Neuigkeiten werden veröffentlicht, sobald wir zurück sind!';

########## Forum Info ##########
/**
 * The name of your forum.
 * @var string
 */
$mbname = 'SMF Beta 4';
/**
 * The default language file set for the forum.
 * @var string
 */
$language = 'german';
/**
 * URL to your forum's folder. (without the trailing /!)
 * @var string
 */
$boardurl = 'xxx';
/**
 * Email address to send emails from. (like noreply@yourdomain.com.)
 * @var string
 */
$webmaster_email = 'xxx';
/**
 * Name of the cookie to set for authentication.
 * @var string
 */
$cookiename = 'SMFCookiexxx';

########## Database Info ##########
/**
 * The database type
 * Default options: mysql, postgresql
 * @var string
 */
$db_type = 'mysql';
/**
 * The server to connect to (or a Unix socket)
 * @var string
 */
$db_server = 'xxx';
/**
 * The database name
 * @var string
 */
$db_name = 'xxx';
/**
 * Database username
 * @var string
 */
$db_user = 'xxx';
/**
 * Database password
 * @var string
 */
$db_passwd = 'xxx';
/**
 * Database user for when connecting with SSI
 * @var string
 */
$ssi_db_user = '';
/**
 * Database password for when connecting with SSI
 * @var string
 */
$ssi_db_passwd = '';
/**
 * A prefix to put in front of your table names.
 * This helps to prevent conflicts
 * @var string
 */
$db_prefix = 'smf_';
/**
 * Use a persistent database connection
 * @var int|bool
 */
$db_persist = 1;
/**
 *
 * @var int|bool
 */
$db_error_send = 1;

########## Cache Info ##########
/**
 * Select a cache system. You want to leave this up to the cache area of the admin panel for
 * proper detection of apc, memcache, output_cache, smf, or xcache
 * (you can add more with a mod).
 * @var string
 */
$cache_accelerator = 'smf';
/**
 * The level at which you would like to cache. Between 0 (off) through 3 (cache a lot).
 * @var int
 */
$cache_enable = 1;
/**
 * This is only used for memcache / memcached. Should be a string of 'server:port,server:port'
 * @var array
 */
$cache_memcached = '';
/**
 * This is only for the 'smf' file cache system. It is the path to the cache directory.
 * It is also recommended that you place this in /tmp/ if you are going to use this.
 * @var string
 */
$cachedir = 'xxx/smfbeta/cache';

########## Image Proxy ##########
# This is done entirely in Settings.php to avoid loading the DB while serving the images
/**
 * Whether the proxy is enabled or not
 * @var bool
 */
$image_proxy_enabled = 0;

/**
 * Secret key to be used by the proxy
 * @var string
 */
$image_proxy_secret = 'xxx';

/**
 * Maximum file size (in KB) for indiviudal files
 * @var int
 */
$image_proxy_maxsize = 5192;

########## Directories/Files ##########
# Note: These directories do not have to be changed unless you move things.
/**
 * The absolute path to the forum's folder. (not just '.'!)
 * @var string
 */
$boarddir = 'xxx/smfbeta';
/**
 * Path to the Sources directory.
 * @var string
 */
$sourcedir = 'xxx/smfbeta/Sources';
/**
 * Path to the Packages directory.
 * @var string
 */
$packagesdir = 'xxx/smfbeta/Packages';
/**
 * Path to the tasks directory.
 * @var string
 */
$tasksdir = 'xxx/smfbeta/Sources/tasks';

# Make sure the paths are correct... at least try to fix them.
if (!file_exists($boarddir) && file_exists(dirname(__FILE__) . '/agreement.txt'))
	$boarddir = dirname(__FILE__);
if (!file_exists($sourcedir) && file_exists($boarddir . '/Sources'))
	$sourcedir = $boarddir . '/Sources';
if (!file_exists($cachedir) && file_exists($boarddir . '/cache'))
	$cachedir = $boarddir . '/cache';

########## Error-Catching ##########
# Note: You shouldn't touch these settings.
if (file_exists((isset($cachedir) ? $cachedir : dirname(__FILE__)) . '/db_last_error.php'))
	include((isset($cachedir) ? $cachedir : dirname(__FILE__)) . '/db_last_error.php');


if (!isset($db_last_error))
{
	// File does not exist so lets try to create it
	file_put_contents((isset($cachedir) ? $cachedir : dirname(__FILE__)) . '/db_last_error.php', '<' . '?' . "php\n" . '$db_last_error = 0;' . "\n" . '?' . '>');
	$db_last_error = 0;
}

$db_character_set = 'utf8';

$db_show_debug = false;

$upgradeData = '';
?>
```

**Comment by wintstar** [comment]
Yes
`global_character_set | UTF-8`

**Comment by wintstar** [comment]
Your Fix #4734   works. You can see it on the test board.
